### ISSUE TITLE
Support las file import/export with normals

### ISSUE THREAD

**Issue by jo-chemla** [CloudCompare/CloudCompare#1848]
**Describe the feature you would like**

Normals can be stored per point for a `las` file using extra dimensions (or Extra Bytes) - different naming conventions exist, see [this thread](https://github.com/ASPRSorg/LAS/issues/37#issuecomment-1695757865). The most used one (and that used by PDAL) seem to be `NormalX` - while other include `nx, normal_x, normalx, normal x`.
It would be nice for CloudCompare to support the normals dimension at las reading/writing. PDAL can already do this - see below. 

- [ ] Support normals writing for las export 
- [ ] Support normals parsing for las import

**Describe alternatives you've considered**
Reading/Writing such fields can be done using `PDAL`. Example PDAL for writing normals within a las file is based on adding the property `"extra_dims": "all"` to the `writers.las` filter. 
All variations of `"extra_dims": "nx=float, NormalY=double, normal_Z=float"` also do work, and they end up in the las file having fields named accoring to the command, but the preferred convention would probably be `NormalX`. 
```json
[
  {
    "type": "readers.ply",
    "filename": "ply-with-normals.ply"
  },
  {
    "type": "writers.las",
    "filename": "las-pdal-with-normals.las",
    "extra_dims": "all",
    "extra_dims_also_working": "nx=float, NormalY=double, normal_Z=float"
  }
]
```
The above PDAL pipeline result in the following extra bytes. 
```bash
C:> LASTools\bin\lasinfo.exe las-pdal-with-normals.las

lasinfo (221128) report for 'las-pdal-with-normals.las'
reporting all LAS header entries:
  file signature:             'LASF'
  version major.minor:        1.2
  system identifier:          'PDAL'
  generating software:        'PDAL 2.5.6 (Releas)'
  header size:                227
  offset to point data:       857
  number of point records:    3269820
  scale factor x y z:         0.01 0.01 0.01
  offset x y z:               0 0 0
  min x y z:                  -120.03 -107.38 -2.27
  max x y z:                  83.05 81.43 66.85
variable length header record 1 of 1:
  user ID              'LASF_Spec'
  length after header  576
  description          'Extra Bytes Record'
    Extra Byte Descriptions
      data type: 9 (float), name "nx", description: "X component of a vector normal t", scale: 1 (not set), offset: 0 (not set)
      data type: 9 (float), name "NormalY", description: "Y component of a vector normal t", scale: 1 (not set), offset: 0 (not set)
      data type: 9 (float), name "normal_Z", description: "Z component of a vector normal t", scale: 1 (not set), offset: 0 (not set)
```

**Additional context**
There does not yet seem to be an agreed upon convention for storing normals within las files, but there are several signs of industrials (Metashape, TerraScan, Riegel) using las extra dimensions (or Extra Bytes) to store normals. It would be great or CloudCompare to support Writing/Reading these normals per vertex. 

Edit: Seems like there are both [qPDALIO](https://github.com/CloudCompare/CloudCompare/tree/master/plugins/core/IO/qPDALIO) based on pdal and [qLASIO](https://github.com/CloudCompare/CloudCompare/tree/master/plugins/core/IO/qLASIO) based on laszip, I imagine the las save dialog uses qLASIO.

**Comment by jo-chemla** [comment]
Note the convention for storing normals within LAS files seem to be `NormalX` according to a post on the convention thread: https://github.com/ASPRSorg/LAS/issues/37#issuecomment-1695757865

**Comment by dgirardeau** [comment]
Hi @tmontaigu. Do you think this is something we could support with the new LAS_IO plugin?

**Comment by jo-chemla** [comment]
Thanks for getting back! And noted for the shift to a custom LAS FILTER not dependent on pdal or liblas. 

Probably a question for Thomas then, are the reasons why you made the switch and decided not to use pdal or liblas listed somewhere (wiki or repo)? I would have thought both libs were mature enough/accepting contributions when there was a need for implementing missing features. 

Exporting a las with normals is handled by pdal by default when point have fields storing normal coords for example. I was a bit quick to interpret the convention for naming normals, there is no standard at the moment although the consensus tends to point towards NormalX, more on the above ASPRS thread.

**Comment by tmontaigu** [comment]
> Probably a question for Thomas then, are the reasons why you made the switch and decided not to use pdal or liblas listed somewhere (wiki or repo)? I would have thought both libs were mature enough/accepting contributions when there was a need for implementing missing features.

Kinda answered here https://github.com/CloudCompare/CloudCompare/issues/1845#issuecomment-1696259722 

- liblas should not be used, it has been deprecated since 2018, it does not support LAS/LAZ 1.4.
- cloudcompare used to used liblas, but it became a problem because it did not (and still does not, and likely never will)
  support LAZ/LAS 1.4, so I replaced it with PDAL https://github.com/CloudCompare/CloudCompare/pull/570
- PDAL does not support LAS with waveforms which we need to support
- We won't want to maintain two LAS plugins (one with waveform support one without)

> Exporting a las with normals is handled by pdal by default when point have fields storing normal coords for example. I was a bit quick to interpret the convention for naming normals, there is no standard at the moment although the consensus tends to point towards NormalX, more on the above ASPRS thread.

PDAL only support exporting normals that are in a PDAL point cloud data structure, In cloud compare point clouds are in cloud compare data structure, so we would still need to write the code that cloud compare's normals into pdal normals.

Storing normals in LAS is done (and can only be done) by storing them as extra bytes, since the new plugin support that we can support it

The lib used in the new plugin is LASZip, the official/original implementation of the LAZ compression

**Comment by jo-chemla** [comment]
Thanks for the link and details. I had PDAL+[lastools](https://github.com/LAStools/LAStools) in mind rather than PDAL+liblas - sorry for the confusion. The [LASZip](https://github.com/LASzip/LASzip) lib you are using in the new CC LAS plugin seem to be part of lastools so thanks for the clarification. Of course understand why it's preferred to maintain a single LAS plugin. 

Note for checking of common extra bytes field names (for example for input LAS extra bytes parsing), you could also use PDAL [Dimensions.json](https://github.com/PDAL/PDAL/blob/master/pdal/Dimension.json) file - even if you are not using PDAL but LASZip.

PS I've never used a las with Full Waveform Data within CC, must be pretty nice.

**Comment by tmontaigu** [comment]
closed by #1882

**Comment by jo-chemla** [comment]
Thanks for the feature!

**Comment by dgirardeau** [comment]
It's normally available in the latest 2.13.beta release (online)
